src/ray/rpc/grpc_server.cc

```diff
@@ -179,13 +184,29 @@ void GrpcServer::PollEventsFromCompletionQueue(int index) {
   // Keep reading events from the `CompletionQueue` until it's shutdown.
   while (cqs_[index]->Next(&tag, &ok)) {
     auto *server_call = static_cast<ServerCall *>(tag);
+
+    const auto &factory = server_call->GetServerCallFactory();
+    // NOTE: For these handlers that have set it to -1, we set default (per
+    //       thread) buffer at 32, though it doesn't have any impact on concurrency
+    //       (since we're recreating new instance of `ServerCall` as soon as one
+    //       gets occupied therefore not serving as back-pressure mechanism)
+    const bool is_unbounded_rpc = factory.GetMaxActiveRPCs() == -1;
+
     bool delete_call = false;
     // A new call is needed after the server sends a reply, no matter the reply is
     // successful or failed.
     bool need_new_call = false;
     if (ok) {
       switch (server_call->GetState()) {
       case ServerCallState::PENDING:
+        if (is_unbounded_rpc) {
+          // Create a new `ServerCall` to accept the next incoming request.
+          // We create this before handling the request only when no back pressure
+          // limit is set. So that the it can be populated by the completion queue
+          // in the background if a new request comes in.
+          factory.CreateCall();
+        }
+
```

Comment:
For this PR, let's revert this change as it's unrelated to `RAY_gcs_max_active_rpcs_per_handler`. There is actually a subtle difference between the old code and the new code: with the old code, a new call is created inside `ServerCall::HandleRequestImpl` which runs in the io service event loop while with the new code, a new call is created inside `GrpcServer::PollEventsFromCompletionQueue` which runs in the polling threads. With the old code, where is some degree of backpressure if the io service event loop is blocked.

Reply:
That's intentional change -- for unbounded RPCs there should be no back-pressure applied by definition

Reply:
This is another place where the code might be "wrong "based on the definition but this is the behavior we have for years so changing it should be careful. We should split it into a separate PR and just testing the impact of this particular change. Currently this PR has two behavior changes.

Reply:
Ack. On the same page.